ci: nix fmt action

dougch opened the pull request:
### Resolved issues:

n/a

### Description of changes: 

Run `nix fmt` and `nix flake check` as a github action and fail if anything is changed or fails a flake check.

### Call-outs:

- Example failing on bad format https://github.com/dougch/s2n-tls/actions/runs/4178332260/jobs/7236991304
- Example failing on bad flake https://github.com/dougch/s2n-tls/actions/runs/4178371463/jobs/7237080931


### Testing:

 How is this change tested (unit tests, fuzz tests, etc.)? fork

 Is this a refactor change? no

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

harrisonkaiser commented on .github/workflows/ci_nix.yml:
```diff
@@ -0,0 +1,18 @@
+        run: nix flake check
+      - name: nix fmt
+        run: |
+          nix fmt $(find . -maxdepth 2 -name \*nix -type f)
```
We probably don't want `-maxdepth` here. I have a draft PR that adds nix files in `codebuild/bin`. They should probably go somewhere else, but no reason to limit the formatter.

dougch replied:
It was slow on my host; removed in the latest commit.  I think the convention is to have a top level `nix` folder, but another PR.

harrisonkaiser reviewed (commented):
This is definitely needed. Thanks for adding it at the beginning -- as we know linters and checks and formatters are relatively more difficult to graft on afterward. 

Should this be part of the `ci_linting.yml` GitHub action? These both feel like linters.

dougch commented:
> 

Agree- moved these changes into ci_linting

camshaft commented:
is there a way to know what file has the bad formatting? from the sample run it just exits without any error message. at the very least I would say it needs to say something like "formatting failed, please run `nix fmt ...`" so people know how to address it.

dougch commented:
> is there a way to know what file has the bad formatting? from the sample run it just exits without any error message. at the very least I would say it needs to say something like "formatting failed, please run `nix fmt ...`" so people know how to address it.

Refactored the fmt check to be more verbose and updated the example failure link above.

harrisonkaiser commented on .github/workflows/ci_linting.yml:
```diff
@@ -90,3 +90,34 @@ jobs:
+        with:
+          nix_conf: experimental-features = nix-command flakes
+      - name: nix flake check
+        run: nix flake check
```
I plan on merging a PR that is only tested on x86_64-linux. Ultimately the goal is to support all systems. Can we add: `NIXPKGS_ALLOW_UNSUPPORTED_SYSTEM=1 nix flake check --impure` for the moment?

Similarly: we are currently depending on Python3.9. The `pyopenssl` package is a test depend that is marked broken in the nix package repo. The goal is to bump our python dependency (`pyopenssl` for Python3.10 is not marked broken). Can we add `NIX_PKGS_ALLOW_BROKEN=1` for the moment?

harrisonkaiser reviewed (changes requested):
Nix flake check may be broken by the upcoming devShell MVP.

dougch commented:
Created a test branch with this change and changes from #3824 to show the `nix flake` check passing https://github.com/dougch/s2n-tls/actions/runs/4237880185/jobs/7364331113

harrisonkaiser reviewed (changes requested):
Remove ci_nix.yml
